### PR TITLE
CMakeLists and C++ files refactoring.

### PR DESCRIPTION
**Pull request by vcarpani**
This PR refactors the CMakeLists file and the C++ files, to make it easier to package the C++ version of this library.

The main problem is due to the usage of `INTERFACE_INCLUDE_DIRECTORIES` in the CMakeLists file, _since it is not advisable to populate the_ `INSTALL_INTERFACE` _of the_ `INTERFACE_INCLUDE_DIRECTORIES` _of a target with paths for dependencies. That would hard-code into installed packages the include directory paths for dependencies as found on the machine the package was made on_ [[source](https://cmake.org/cmake/help/v3.2/prop_tgt/INTERFACE_INCLUDE_DIRECTORIES.html)].

Changes:
- Moved cpp headers to `cinclude` folder.
- Refactor cpp source files to account for the move of headers file.
- Use cmake `include_directories` instead of `INTERFACE_INCLUDE_DIRECTORIES`

There is still some work to be done on the ROCM version of the python library.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.0)
 project(torchvision)
 set(CMAKE_CXX_STANDARD 14)
-set(TORCHVISION_VERSION 0.6.0)
+set(TORCHVISION_VERSION 0.5.0)
```
Comment (fmassa) [inline CMakeLists.txt:4]: this seems like an unwanted change?

Comment (vcarpani) [reply CMakeLists.txt:4]: Yes, sorry for the typo

**Comment by vcarpani** [comment]
> Hi,
> 
> Thanks for the PR!
> 
> Just so that I understand a bit better, in this PR do you target the models or the operators most specifically?
> I don't have much experience with CMake, so I won't be able to review this PR. Also, how does it compare to #2253 ?
> 
> Maybe @bmanga could have a look?

Hi, thanks for the quick feedback!

This PR targets all the C++ code, a part from `decoder`, `image` and `video_reader`. It is not directly related with #2253, but if #2253 gets merged before this, this PR is going to need some more work or viceversa.

**Comment by vcarpani** [comment]
Any news on this?

**Comment by bmanga** [comment]
Hi @vcarpani, can you edit the PR message to list all the changes you made? Thanks!

File: CMakeLists.txt [hunk 2]
```diff
@@ -11,64 +11,70 @@ if(WITH_CUDA)
...
+    EXPORT TorchVisionTargets
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
 
 install(EXPORT TorchVisionTargets
-  NAMESPACE TorchVision::
-  DESTINATION ${TORCHVISION_CMAKECONFIG_INSTALL_DIR})
+    NAMESPACE TorchVision::
+    DESTINATION ${TORCHVISION_CMAKECONFIG_INSTALL_DIR}
+)
```
Comment (bmanga) [inline CMakeLists.txt:66]: These all look like unrelated formatting changes. I'd say stick to the current style, and if you want to propose a different one do so in a different PR.

**Comment by vcarpani** [comment]
Hi @bmanga, last commit should solve your comments, plus I added the list of changes to the description of the PR

**Comment by fmassa** [comment]
Naive question: would it be possible to avoid moving the header files into a new directory?

**Comment by vcarpani** [comment]
> Naive question: would it be possible to avoid moving the header files into a new directory and concentrate the changes only into CMakeLists?

From my knowledge of CMake not, but I could look into that somewhere this week

**Comment by fmassa** [comment]
The reason why I'm asking is because this change will require some internal changes as well to torchvision, and in PyTorch they do [mix headers and implementation in the same folder](https://github.com/pytorch/pytorch/tree/master/aten/src/ATen), so I would say I would prefer if we could keep it that way.

**Comment by bmanga** [comment]
@vcarpani have you tried to `include_directories(torchvision/csrc)`?

**Comment by vcarpani** [comment]
@bmanga that would probably work (it is also the same approach the pytorch uses), but it is also quite horrible, since you would also get the cpp files in the include directories

**Comment by bmanga** [comment]
@vcarpani if you mean during installation, you can specify a PATTERN for cmake's `install` command.

**Comment by vcarpani** [comment]
@bmanga, yes I was referring to that, since it forces you to specify the installation pattern, while it is much simpler if you have only `.h` or `.hpp` files in the `include_directories`

**Comment by bmanga** [comment]
@vcarpani I understand your point and I agree in principle, however I don't think it's a worthwhile change as of now. I'd suggest to go for the `include_directories(torchvision/csrc)` + install pattern for this PR. It should address the problem with `INTERFACE_INCLUDE_DIRECTORIES` and be minimally disruptive.

**Comment by fmassa** [comment]
I agree with @bmanga comment, @vcarpani could you change your PR to do what @bmanga proposed?
